internal/parse/parser.go

```diff
@@ -187,57 +209,143 @@ func isNameByte(c byte) bool {
-		{bytef: p.skipByte, result: []bool{false}, input: "abc", data: []string{"b"}},
-		{bytef: p.skipByte, result: []bool{true, true}, input: "abc", data: []string{"a", "b"}},
+	// Case 1: A single column.
+	if col, ok := p.parseFullName(columnId); ok {
+		cols = append(cols, col)
 
-		{bytef: p.skipByteFind, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.skipByteFind, result: []bool{false, true, true}, input: "abcde", data: []string{"x", "b", "c"}},
-		{bytef: p.skipByteFind, result: []bool{true, false}, input: "abcde ", data: []string{" ", " "}},
+		// Case 2: Multiple columns.
+	} else if p.skipByte('(') {
+		col, ok := p.parseFullName(columnId)
+		cols = append(cols, col)
+		p.skipSpaces()
+		// If the column names are not formated in a recognisable way then give
+		// up trying to parse.
+		if !ok {
+			return cols, false
+		}
+		for p.skipByte(',') {
+			p.skipSpaces()
+			col, ok := p.parseFullName(columnId)
+			p.skipSpaces()
+			if !ok {
+				return cols, false
+			}
+			cols = append(cols, col)
+		}
+		p.skipSpaces()
+		p.skipByte(')')
+	}
+	p.skipSpaces()
+	return cols, true
+}
 
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "abc    d", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{true}, input: "     abcd", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{true}, input: "  \t  abcd", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "\t  abcd", data: []string{}},
+// parseInputExpression parses an SDL input go-defined type to be used as a
+// query argument.
+func (p *Parser) parseInputExpression() (*InputPart, bool, error) {
+	cp := p.save()
+	var err error
+	var fn FullName
+	var ok bool
 
-		{stringf: p.skipString, result: []bool{false}, input: "", data: []string{"a"}},
-		{stringf: p.skipString, result: []bool{true, true}, input: "helloworld", data: []string{"hElLo", "w"}},
-		{stringf: p.skipString, result: []bool{true, true}, input: "hello world", data: []string{"hello", " "}},
+	p.skipSpaces()
+	if p.skipByte('$') {
+		fn, ok = p.parseFullName(typeId)
+		if !ok {
+			err = fmt.Errorf("malformed input type")
+		}
+		p.skipSpaces()
+		return &InputPart{fn}, true, err
 	}
-	for _, v := range parseTests {
-		p.Parse(v.input)
-		for i, _ := range v.result {
-			var result bool
-			if v.bytef != nil {
-				result = v.bytef(v.data[i][0])
-			}
-			if v.stringf != nil {
-				result = v.stringf(v.data[i])
-			}
-			if v.stringf0 != nil {
-				result = v.stringf0()
-			}
-			if v.result[i] != result {
-				log.Printf("Test %#v failed. Expected: '%t', got '%t'\n", v, result, v.result[i])
+	cp.restore()
+	return nil, false, nil
+}
+
+// parseInputExpression parses an SDL input go-defined type to be used as a
+// query argument.
+func (p *Parser) parseStringLiteral() (*BypassPart, bool, error) {
+	cp := p.save()
+	p.skipSpaces()
```

Comment:
Per above, these skipSpaces seems a bit problematic as it prevents detecting cases where spaces are not present. The outer layer is probably the right place to be doing the skipping.

Reply:
skipSpaces removed.